cpp/src/centrality/betweenness_centrality_impl.cuh

```diff
@@ -15,46 +15,368 @@
+
+  rmm::device_uvector<vertex_t> sigma(graph_view.local_vertex_partition_range_size(),
+                                      handle.get_stream());
+  rmm::device_uvector<vertex_t> distance(graph_view.local_vertex_partition_range_size(),
+                                         handle.get_stream());
+  detail::scalar_fill(handle, distance.data(), distance.size(), invalid_distance);
+  detail::scalar_fill(handle, sigma.data(), sigma.size(), vertex_t{0});
+
+  edge_src_property_t<graph_view_t<vertex_t, edge_t, false, multi_gpu>, vertex_t> src_sigma(
+    handle, graph_view);
+  edge_dst_property_t<graph_view_t<vertex_t, edge_t, false, multi_gpu>, vertex_t> dst_distance(
+    handle, graph_view);
+
+  auto vertex_partition =
+    vertex_partition_device_view_t<vertex_t, multi_gpu>(graph_view.local_vertex_partition_view());
+
+  if (vertex_frontier.bucket(bucket_idx_cur).size() > 0) {
+    thrust::for_each(
+      handle.get_thrust_policy(),
+      vertex_frontier.bucket(bucket_idx_cur).begin(),
+      vertex_frontier.bucket(bucket_idx_cur).end(),
+      [d_sigma = sigma.begin(), d_distance = distance.begin(), vertex_partition] __device__(
+        auto v) {
+        auto offset        = vertex_partition.local_vertex_partition_offset_from_vertex_nocheck(v);
+        d_distance[offset] = 0;
+        d_sigma[offset]    = 1;
+      });
+  }
+
+  edge_t hop{0};
+
+  while (true) {
+    update_edge_src_property(handle, graph_view, sigma.begin(), src_sigma);
+    update_edge_dst_property(handle, graph_view, distance.begin(), dst_distance);
+
+    auto [new_frontier, new_sigma] =
+      transform_reduce_v_frontier_outgoing_e_by_dst(handle,
+                                                    graph_view,
+                                                    vertex_frontier.bucket(bucket_idx_cur),
+                                                    src_sigma.view(),
+                                                    dst_distance.view(),
+                                                    cugraph::edge_dummy_property_t{}.view(),
+                                                    brandes_e_op_t<vertex_t>{},
+                                                    reduce_op::plus<edge_t>());
+
+    update_v_frontier(handle,
+                      graph_view,
+                      std::move(new_frontier),
+                      std::move(new_sigma),
+                      vertex_frontier,
+                      std::vector<size_t>{bucket_idx_next},
+                      thrust::make_zip_iterator(distance.begin(), sigma.begin()),
+                      thrust::make_zip_iterator(distance.begin(), sigma.begin()),
+                      [hop] __device__(auto v, auto old_values, auto v_sigma) {
+                        vertex_t next_distance = thrust::get<0>(old_values) == invalid_distance
+                                                   ? (hop + 1)
+                                                   : thrust::get<0>(old_values);
```

Comment:
Is this check necessary? We're accumulating dst only when dst_distance == invalid_distance. Did you find any case thrust::get<0>(old_values) != invalid_distance?